drivers/adc/adc_gecko.c

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (c) 2019 Interay Solutions B.V.
+ * Copyright (c) 2019 Oane Kingma
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+
+#include <adc.h>
+#include <device.h>
+#include <kernel.h>
+#include <init.h>
+#include <soc.h>
+#include <em_adc.h>
+#include <em_cmu.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_gecko);
+
+#define ADC_PREFIX cmuClock_ADC
+#define CLOCK_ID_PRFX2(prefix, suffix) prefix##suffix
+#define CLOCK_ID_PRFX(prefix, suffix) CLOCK_ID_PRFX2(prefix, suffix)
+#define CLOCK_ADC(id) CLOCK_ID_PRFX(ADC_PREFIX, id)
+
+/* Acquisition times */
+static const u16_t acq_time_single_tbl[10] = {
+1, 2, 3, 4, 8, 16, 32, 64, 128, 256 };
+
+/* 16 external channels. */
+#define GECKO_CHANNEL_COUNT             32
+
+struct adc_gecko_ch_cfg {
+	ADC_Ref_TypeDef reference;
+	ADC_AcqTime_TypeDef acq_time;
+	ADC_PosSel_TypeDef input_positive;
+};
+
+struct adc_gecko_data {
+	struct adc_context ctx;
+	struct device *dev;
+	u16_t *buffer;
+	u16_t *repeat_buffer;
+	u8_t channel_count;
+	struct adc_gecko_ch_cfg ch_cfg[GECKO_CHANNEL_COUNT];
+};
+
+struct adc_gecko_cfg {
+	ADC_TypeDef *base;
+	CMU_Clock_TypeDef clock;
+	u32_t prescaler;
+	void (*irq_cfg_func)(void);
+};
+
+#define DEV_DATA(dev) \
+	((struct adc_gecko_data *)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((struct adc_gecko_cfg *)(dev)->config->config_info)
+#define CH_CFG(ch) \
+	((struct adc_gecko_ch_cfg *)&data->ch_cfg[ch])
+
+static int adc_gecko_check_buffer_size(const struct adc_sequence *sequence,
+				       u8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(u16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+			sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_gecko_check_resolution(const struct adc_sequence *sequence)
+{
+	int value = 0;
+
+	switch (sequence->resolution) {
+	case 6:
+		value = adcRes6Bit;
+		break;
+	case 8:
+		value = adcRes8Bit;
+		break;
+	case 12:
+		value = adcRes12Bit;
+		break;
+	default:
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	return value;
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	const struct adc_gecko_cfg *config = DEV_CFG(dev);
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+	ADC_InitSingle_TypeDef sInit = ADC_INITSINGLE_DEFAULT;
+	struct adc_gecko_ch_cfg *config_ch = NULL;
+	u8_t index;
+	int res;
+
+	/* Find designated channel, only 1 supported now */
```

Comment:
You should then return an error if more than one channel is selected (or when none is selected, currently you'll get index `0xff` for this case).
It would be also good to ensure that the selected channel has been configured.

Reply:
I have added a check that will return `-EINVAL` if more than 1 channel is enabled. If what you say is correct that index `0xFF` is used when no channel is selected, that check will also return `-EINVAL`. 

Furthermore, I added a boolean `initialized` to each channel configuration to verify that `adc_gecko_channel_setup()` has been completed for that channel before being able to call `start_read()`. 

Please let me know if you think that is an acceptable solution.

Reply:
> If what you say is correct that index `0xFF` is used when no channel is selected

That's because you use `index = find_lsb_set(sequence->channels) - 1;` and `find_lsb_set` returns `0` when its parameter is `0` (i.e. no channel is selected), and `index` is of `u8_t` type.

Reply:
> Please let me know if you think that is an acceptable solution.

The changes you mention look fine.
But would it be actually that hard to provide the support for multiple channels? You could take the pattern from some other driver, [adc_mcux_adc16.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_mcux_adc16.c#L186) for instance.